### PR TITLE
Fix for Workgroup stabilization failures on operations

### PR DESCRIPTION
**Pull request by vramanet**
Description: This CR contains fix for stabilization not correctly implemented for Workgroup Resource. During Read handler we are returing in Progress status which is not allowed. This CR will fix that issue.

Currently as part of create and delete handlers we are only checking if Workgroup is stablized. This is causing failures if the same workgroup is created immediately causing "Operation in progress failure". This is because Namespace is not stabilized during the operation. This CR fixes that issue as well.

Testing:
1. Tested if CTV2 is passing in eu-north-1 and us-west-1 region which are currently failing CTV2 for Workgroup resource.
2. Verified CTV2 is passing for those regions in personal account.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/CreateHandler.java [hunk 1]
```diff
@@ -34,73 +40,92 @@ protected ProgressEvent<ResourceModel, CallbackContext> handleRequest(
...
+                .build();
+
+        GetNamespaceResponse getNamespaceResponse = this.readNamespace(getNamespaceRequest, proxyClient);
+
+        logger.log(String.format("%s : Namespace: %s has successfully been read.",
+                ResourceModel.TYPE_NAME, getNamespaceResponse.namespace().namespaceName()));
+
+        logger.log(getNamespaceResponse.toString());
+
+        return NAMESPACE_STATUS_AVAILABLE.equalsIgnoreCase(getNamespaceResponse.namespace().statusAsString());
```
Comment (mashuaihh) [inline aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/CreateHandler.java:116]: looks like we were checking workgroup using Enum, which seems safer to me, can we do something like that here, or why do we need equalsIgnoreCase instead?

File: aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/CreateHandler.java [hunk 2]
```diff
@@ -34,73 +40,92 @@ protected ProgressEvent<ResourceModel, CallbackContext> handleRequest(
...
     }
 
-    private ProgressEvent<ResourceModel, CallbackContext> createWorkgroupErrorHandler(final CreateWorkgroupRequest awsRequest,
+    private ProgressEvent<ResourceModel, CallbackContext> createWorkgroupErrorHandler(final Object awsRequest,
```
Comment (mashuaihh) [inline aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/CreateHandler.java:119]: what's the reason for this change?

File: aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/DeleteHandler.java [hunk 3]
```diff
@@ -31,43 +35,94 @@ protected ProgressEvent<ResourceModel, CallbackContext> handleRequest(
...
+
+            logger.log(getWorkgroupResponse.toString());
+        } catch (ResourceNotFoundException e) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private boolean isNamespaceStable(final Object awsRequest,
```
Comment (mashuaihh) [inline aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/DeleteHandler.java:97]: do we need awsRequest here?

File: aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/UpdateHandler.java [hunk 4]
```diff
@@ -113,13 +123,24 @@ private ResourceModel getUpdatableResourceModel(ResourceModel desiredModel, Reso
                 .build();
     }
 
-    private GetWorkgroupResponse readWorkgroup(final GetWorkgroupRequest awsRequest,
-                                               final ProxyClient<RedshiftServerlessClient> proxyClient) {
-        GetWorkgroupResponse awsResponse;
-        awsResponse = proxyClient.injectCredentialsAndInvokeV2(awsRequest, proxyClient.client()::getWorkgroup);
+    private boolean isWorkgroupStable(final Object awsRequest,
```
Comment (mashuaihh) [inline aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/UpdateHandler.java:126]: same here, let's remove the unused param awsRequest?

Comment (mashuaihh) [inline aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/UpdateHandler.java:126]: we are recreating this method multiple times, can we move them to the base class and use only one?

File: aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/UpdateHandler.java [hunk 5]
```diff
@@ -154,82 +175,14 @@ private TagResourceResponse updateTags(final UpdateTagsRequest awsRequest,
...
-            } catch (InterruptedException ie) {
-                Thread.currentThread().interrupt(); // Restore the interrupted status
-                throw new RuntimeException("Interrupted during retry wait", ie);
-            }
-        }
-
-    }
-
-    private ProgressEvent<ResourceModel, CallbackContext> updateWorkgroupErrorHandler(final UpdateWorkgroupRequest awsRequest,
+    private ProgressEvent<ResourceModel, CallbackContext> updateWorkgroupErrorHandler(final Object awsRequest,
```
Comment (mashuaihh) [inline aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/UpdateHandler.java:178]: keep the type?

File: aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/CreateHandler.java (hunk 2 above)
Comment (vramanet) [reply aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/CreateHandler.java:119]: There are multiple stabilization calls which are using the same error handler so aws request makes sense here. We are logging the exception operation in logs so we should be good here.

File: aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/UpdateHandler.java (hunk 5 above)
Comment (vramanet) [reply aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/UpdateHandler.java:178]: There are multiple stabilization calls which are using the same error handler so aws request makes sense here. We are logging the exception operation in logs so we should be good here.

File: aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/UpdateHandler.java (hunk 4 above)
Comment (vramanet) [reply aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/UpdateHandler.java:126]: isWorkgroupStable check is performed after multiple steps in update handler i.e during update tags, update workgroup instance so it makes more sense to use object class here

Comment (vramanet) [reply aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/UpdateHandler.java:126]: Will do

File: aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/DeleteHandler.java (hunk 3 above)
Comment (vramanet) [reply aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/DeleteHandler.java:97]: Will update.

File: aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/BaseHandlerStd.java [hunk 6]
```diff
@@ -23,12 +36,19 @@ public abstract class BaseHandlerStd extends BaseHandler<CallbackContext> {
     public static final String BUSY_WORKGROUP_RETRY_EXCEPTION_MESSAGE =
             "There is an operation running on the existing workgroup";
 
+    protected static final String NAMESPACE_STATUS_AVAILABLE = "available";
```
Comment (hkondru) [inline aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/BaseHandlerStd.java:39]: Can't we use NamespaceStatus enum instead?

Comment (vramanet) [reply aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/BaseHandlerStd.java:39]: This is not being used will remove this.

File: aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/DeleteHandler.java [hunk 7]
```diff
@@ -31,43 +36,94 @@ protected ProgressEvent<ResourceModel, CallbackContext> handleRequest(
...
+                                .makeServiceCall((awsRequest, sdkProxyClient) -> {
+                                    GetNamespaceResponse awsResponse = this.readNamespace(awsRequest, sdkProxyClient);
+
+                                    logger.log(String.format("%s : %s has successfully been created.",
+                                            ResourceModel.TYPE_NAME, awsResponse.namespace().namespaceName()));
+                                    logger.log(awsResponse.toString());
 
+                                    return awsResponse;
+                                })
+                                .stabilize(this::isNamespaceStable)
```
Comment (hkondru) [inline aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/DeleteHandler.java:67]: We need to check if the workgroup is stable or not before deleting the workgroup. Namespace stability check is not required

File: aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/CreateHandler.java (hunk 2 above)
Comment (mashuaihh) [reply aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/CreateHandler.java:120]: still don't understand, we're using createWorkgroupErrorHandler everywhere?

File: aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/DeleteHandler.java [hunk 8]
```diff
@@ -31,43 +36,94 @@ protected ProgressEvent<ResourceModel, CallbackContext> handleRequest(
...
+
+            logger.log(getWorkgroupResponse.toString());
+        } catch (ResourceNotFoundException e) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private boolean isNamespaceStable(final GetNamespaceRequest awsRequest,
```
Comment (hkondru) [inline aws-redshiftserverless-workgroup/src/main/java/software/amazon/redshiftserverless/workgroup/DeleteHandler.java:98]: Can we add the methods isNamespaceStable, isWorkgroupStable in a base class?

**Review by mashuaihh — APPROVED** [review]

**Review by hkondru — APPROVED** [review]
